server/ast/expr.go

```diff
@@ -0,0 +1,517 @@
+	case *tree.NumVal:
+		switch node.Kind() {
+		case constant.Int:
+			return vitess.NewIntVal([]byte(node.FormattedString())), nil
+		case constant.Float:
+			return vitess.NewFloatVal([]byte(node.FormattedString())), nil
+		default:
+			return nil, fmt.Errorf("unknown number format")
+		}
+	case *tree.OrExpr:
+		left, err := nodeExpr(node.Left)
+		if err != nil {
+			return nil, err
+		}
+		right, err := nodeExpr(node.Right)
+		if err != nil {
+			return nil, err
+		}
+		return &vitess.OrExpr{
+			Left:  left,
+			Right: right,
+		}, nil
+	case *tree.ParenExpr:
+		expr, err := nodeExpr(node.Expr)
+		if err != nil {
+			return nil, err
+		}
+		return &vitess.ParenExpr{
+			Expr: expr,
+		}, nil
+	case *tree.PartitionMaxVal:
+		return nil, fmt.Errorf("MAX is not yet supported")
+	case *tree.PartitionMinVal:
+		return nil, fmt.Errorf("MIN is not yet supported")
+	case *tree.Placeholder:
+		//TODO: figure out if I can delete this
+		return nil, fmt.Errorf("this should probably be deleted (internal error, Placeholder)")
+	case *tree.RangeCond:
+		operator := vitess.BetweenStr
+		if node.Not {
+			operator = vitess.NotBetweenStr
+		}
+		left, err := nodeExpr(node.Left)
+		if err != nil {
+			return nil, err
+		}
+		from, err := nodeExpr(node.From)
+		if err != nil {
+			return nil, err
+		}
+		to, err := nodeExpr(node.To)
+		if err != nil {
+			return nil, err
+		}
+		if node.Symmetric {
+			return &vitess.OrExpr{
+				Left: &vitess.RangeCond{
+					Operator: operator,
+					Left:     left,
+					From:     from,
+					To:       to,
+				},
+				Right: &vitess.RangeCond{
+					Operator: operator,
+					Left:     left,
+					From:     to,
+					To:       from,
+				},
+			}, nil
+		} else {
+			return &vitess.RangeCond{
+				Operator: operator,
+				Left:     left,
+				From:     from,
+				To:       to,
+			}, nil
+		}
+	case *tree.StrVal:
+		//TODO: determine what to do when node.WasScannedAsBytes() is true
+		return &vitess.SQLVal{
+			Type: vitess.StrVal,
+			Val:  []byte(node.RawString()),
+		}, nil
+	case *tree.Subquery:
+		return nodeSubquery(node)
+	case *tree.Tuple:
+		return nil, fmt.Errorf("tuples are not yet supported")
+	case *tree.TupleStar:
+		return nil, fmt.Errorf("(E).* is not yet supported")
+	case *tree.UnaryExpr:
+		expr, err := nodeExpr(node.Expr)
+		if err != nil {
+			return nil, err
+		}
+		switch node.Operator {
+		case tree.UnaryMinus:
+			return &vitess.UnaryExpr{
+				Operator: vitess.UMinusStr,
+				Expr:     expr,
+			}, nil
+		case tree.UnaryComplement:
+			return &vitess.UnaryExpr{
+				Operator: vitess.TildaStr,
+				Expr:     expr,
+			}, nil
+		case tree.UnarySqrt:
+			//TODO: replace with a function
+			return nil, fmt.Errorf("square root operator is not yet supported")
+		case tree.UnaryCbrt:
+			//TODO: replace with a function
+			return nil, fmt.Errorf("cube root operator is not yet supported")
+		default:
+			return nil, fmt.Errorf("the unary operator used is not yet supported")
```

Comment:
This seems impossible no?>

Reply:
Nope! Not impossible, maybe unlikely, but definitely not impossible.